"Error: Cannot find module" output should include filename

When using browserify and there's a problem with the path to a module, I get error output like:

```
Error: Cannot find module 'module' from '/dir/subdir'
```

There may be any number of files in `subdir` that were trying to require the module. It would be really helpful for the error output to include the filename that contained the `require()` call instead of just the path to the dir it resides in.

--- a/lib/async.js
+++ b/lib/async.js
@@ -33,21 +33,22 @@ module.exports = function resolve(x, options, callback) {
 
     var extensions = opts.extensions || ['.js'];
     var basedir = opts.basedir || path.dirname(caller());
+    var parent = opts.filename || basedir;
 
     opts.paths = opts.paths || [];
 
     if (/^(?:\.\.?(?:\/|$)|\/|([A-Za-z]:)?[/\\])/.test(x)) {
-        var res = path.resolve(basedir, x);
+        var res = path.resolve(parent, x);
         if (x === '..' || x.slice(-1) === '/') res += '/';
-        if (/\/$/.test(x) && res === basedir) {
+        if (/\/$/.test(x) && res === parent) {
             loadAsDirectory(res, opts.package, onfile);
         } else loadAsFile(res, opts.package, onfile);
-    } else loadNodeModules(x, basedir, function (err, n, pkg) {
+    } else loadNodeModules(x, parent, function (err, n, pkg) {
         if (err) cb(err);
         else if (core[x]) return cb(null, x);
         else if (n) return cb(null, n, pkg);
         else {
-            var moduleError = new Error("Cannot find module '" + x + "' from '" + basedir + "'");
+            var moduleError = new Error("Cannot find module '" + x + "' from '" + parent + "'");
             moduleError.code = 'MODULE_NOT_FOUND';
             cb(moduleError);
         }
@@ -60,7 +61,7 @@ module.exports = function resolve(x, options, callback) {
             if (err) cb(err);
             else if (d) cb(null, d, pkg);
             else {
-                var moduleError = new Error("Cannot find module '" + x + "' from '" + basedir + "'");
+                var moduleError = new Error("Cannot find module '" + x + "' from '" + parent + "'");
                 moduleError.code = 'MODULE_NOT_FOUND';
                 cb(moduleError);
             }

--- a/lib/sync.js
+++ b/lib/sync.js
@@ -24,24 +24,25 @@ module.exports = function (x, options) {
 
     var extensions = opts.extensions || ['.js'];
     var basedir = opts.basedir || path.dirname(caller());
+    var parent = opts.filename || basedir;
 
     opts.paths = opts.paths || [];
 
     if (/^(?:\.\.?(?:\/|$)|\/|([A-Za-z]:)?[/\\])/.test(x)) {
-        var res = path.resolve(basedir, x);
+        var res = path.resolve(parent, x);
         if (x === '..' || x.slice(-1) === '/') res += '/';
         var m = loadAsFileSync(res) || loadAsDirectorySync(res);
         if (m) return m;
     } else if (core[x]) {
         return x;
     } else {
-        var n = loadNodeModulesSync(x, basedir);
+        var n = loadNodeModulesSync(x, parent);
         if (n) return n;
     }
 
     if (core[x]) return x;
 
-    var err = new Error("Cannot find module '" + x + "' from '" + basedir + "'");
+    var err = new Error("Cannot find module '" + x + "' from '" + parent + "'");
     err.code = 'MODULE_NOT_FOUND';
     throw err;
 

--- a/test/resolver.js
+++ b/test/resolver.js
@@ -3,7 +3,7 @@ var test = require('tape');
 var resolve = require('../');
 
 test('async foo', function (t) {
-    t.plan(10);
+    t.plan(11);
     var dir = path.join(__dirname, 'resolver');
 
     resolve('./foo', { basedir: dir }, function (err, res, pkg) {
@@ -181,7 +181,7 @@ test('normalize', function (t) {
 });
 
 test('cup', function (t) {
-    t.plan(4);
+    t.plan(5);
     var dir = path.join(__dirname, 'resolver');
 
     resolve('./cup', { basedir: dir, extensions: ['.js', '.coffee'] }, function (err, res) {
@@ -200,8 +200,7 @@ test('cup', function (t) {
     });
 
     // Test that filename is reported as the "from" value when passed.
-    resolve('./cup', { basedir: dir, extensions: [ '.js' ], filename: path.join(dir, 'cupboard.js') },
-    function (err, res) {
+    resolve('./cup', { basedir: dir, extensions: ['.js'], filename: path.join(dir, 'cupboard.js') }, function (err, res) {
         t.equal(err.message, "Cannot find module './cup' from '" + path.join(dir, 'cupboard.js') + "'");
     });
 });

--- a/test/resolver.js
+++ b/test/resolver.js
@@ -34,6 +34,11 @@ test('async foo', function (t) {
         t.equal(err.message, "Cannot find module 'foo' from '" + path.resolve(dir) + "'");
         t.equal(err.code, 'MODULE_NOT_FOUND');
     });
+
+    // Test that filename is reported as the "from" value when passed.
+    resolve('foo', { basedir: dir, filename: path.join(dir, 'baz.js') }, function (err) {
+        t.equal(err.message, "Cannot find module 'foo' from '" + path.join(dir, 'baz.js') + "'");
+    });
 });
 
 test('bar', function (t) {
@@ -192,6 +197,12 @@ test('cup', function (t) {
     resolve('./cup', { basedir: dir, extensions: ['.js'] }, function (err, res) {
         t.equal(err.message, "Cannot find module './cup' from '" + path.resolve(dir) + "'");
         t.equal(err.code, 'MODULE_NOT_FOUND');
+    });
+
+    // Test that filename is reported as the "from" value when passed.
+    resolve('./cup', { basedir: dir, extensions: [ '.js' ], filename: path.join(dir, 'cupboard.js') },
+    function (err, res) {
+        t.equal(err.message, "Cannot find module './cup' from '" + path.join(dir, 'cupboard.js') + "'");
     });
 });
 

--- a/test/resolver_sync.js
+++ b/test/resolver_sync.js
@@ -1,7 +1,6 @@
 var path = require('path');
 var test = require('tape');
 var resolve = require('../');
-var path = require('path');
 
 test('foo', function (t) {
     var dir = path.join(__dirname, 'resolver');

--- a/test/resolver_sync.js
+++ b/test/resolver_sync.js
@@ -1,6 +1,7 @@
 var path = require('path');
 var test = require('tape');
 var resolve = require('../');
+var path = require('path');
 
 test('foo', function (t) {
     var dir = path.join(__dirname, 'resolver');
@@ -18,6 +19,17 @@ test('foo', function (t) {
     t.throws(function () {
         resolve.sync('foo', { basedir: dir });
     });
+
+    // Test that filename is reported as the "from" value when passed.
+    t.throws(
+        function () {
+            resolve.sync('foo', { basedir: dir, filename: path.join(dir, 'bar.js') });
+        },
+        {
+            name: 'Error',
+            message: "Cannot find module 'foo' from '" + path.join(dir, 'bar.js') + "'"
+        }
+    );
 
     t.end();
 });

